flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriterFactory.java

```diff
@@ -69,8 +71,16 @@ public RowDataTaskWriterFactory(Table table,
 
     if (equalityFieldIds == null || equalityFieldIds.isEmpty()) {
       this.appenderFactory = new FlinkAppenderFactory(schema, flinkSchema, table.properties(), spec);
+    } else if (upsert) {
+      // In upsert mode, only the new row is emitted using INSERT row kind. Therefore, the deletion schema
+      // must be only the equality fields for that the writer can use the new row to issue a delete by key.
+      //
+      // Outside the upsert case, the entire deleted row is emitted as UPDATE_BEFORE or DELETE row kind,
+      // so the deletion schema must use the whole rows schema in case elements are deleted based on additional
+      // fields.
+      this.appenderFactory = new FlinkAppenderFactory(schema, flinkSchema, table.properties(), spec,
+          ArrayUtil.toIntArray(equalityFieldIds), TypeUtil.select(schema, Sets.newHashSet(equalityFieldIds)), null);
     } else {
-      // TODO provide the ability to customize the equality-delete row schema.
```

Comment:
This comment seems worth removing, as not having the right equality-delete row schema can cause correctness issues.